pkg/reconciler/pipelinerun/affinity_assistant.go

```diff
@@ -40,86 +41,92 @@ import (
-	if (aaBehavior == aa.AffinityAssistantPerWorkspace || aaBehavior == aa.AffinityAssistantDisabled) && pr.HasVolumeClaimTemplate() {
-		if err := c.pvcHandler.CreatePVCsForWorkspaces(ctx, pr.Spec.Workspaces, *kmeta.NewControllerRef(pr), pr.Namespace); err != nil {
-			return fmt.Errorf("failed to create PVC for PipelineRun %s: %w", pr.Name, err)
+			claimTemplateToWorkspace[claimTemplate] = w
 		}
 	}
 	switch aaBehavior {
 	case aa.AffinityAssistantPerWorkspace:
-		for claim, workspaceName := range claimToWorkspace {
+		for claimName, workspaceName := range claimNameToWorkspaceName {
 			aaName := GetAffinityAssistantName(workspaceName, pr.Name)
-			err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []corev1.PersistentVolumeClaimVolumeSource{*claim}, unschedulableNodes)
-			errs = append(errs, err...)
+			if err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []string{claimName}, unschedulableNodes); err != nil {
+				return fmt.Errorf("%w: %v", ErrAffinityAssistantCreationFailed, err)
+			}
 		}
-		for claimTemplate, workspaceName := range claimTemplatesToWorkspace {
-			aaName := GetAffinityAssistantName(workspaceName, pr.Name)
-			// To support PVC auto deletion at pipelinerun deletion time, the OwnerReference of the PVCs should be set to the owning pipelinerun.
-			// In AffinityAssistantPerWorkspace mode, the reconciler has created PVCs (owned by pipelinerun) from pipelinerun's VolumeClaimTemplate at this point,
-			// so the VolumeClaimTemplates are pass in as PVCs when creating affinity assistant StatefulSet for volume scheduling.
-			// If passed in as VolumeClaimTemplates, the PVCs are owned by Affinity Assistant StatefulSet instead of the pipelinerun.
-			err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []corev1.PersistentVolumeClaimVolumeSource{{ClaimName: claimTemplate.Name}}, unschedulableNodes)
-			errs = append(errs, err...)
+		for claimTemplate, workspace := range claimTemplateToWorkspace {
+			// To support PVC auto deletion at pipelinerun deletion time, the OwnerReference of the PVCs should be set to the owning pipelinerun instead of the StatefulSets,
+			// so we create PVCs from PipelineRuns' VolumeClaimTemplate and pass the PVCs to the Affinity Assistant StatefulSet for volume scheduling.
+			if err := c.pvcHandler.CreatePVCFromVolumeClaimTemplate(ctx, workspace, *kmeta.NewControllerRef(pr), pr.Namespace); err != nil {
+				return fmt.Errorf("%w: %v", ErrPvcCreationFailed, err) //nolint:errorlint
+			}
+			aaName := GetAffinityAssistantName(workspace.Name, pr.Name)
+			if err := c.createOrUpdateAffinityAssistant(ctx, aaName, pr, nil, []string{claimTemplate.Name}, unschedulableNodes); err != nil {
+				return fmt.Errorf("%w: %v", ErrAffinityAssistantCreationFailed, err)
+			}
 		}
 	case aa.AffinityAssistantPerPipelineRun, aa.AffinityAssistantPerPipelineRunWithIsolation:
-		if claims != nil || claimTemplates != nil {
+		if claimNames != nil || claimTemplates != nil {
```

Comment:
nit: I suggest `if len(claimNames) != 0` and likewise for claimTemplates, since what you have here differentiates between a nil slice and empty slice which I don't think is intentional.

Reply:
Thanks, fixed